### ISSUE TITLE
WSL2: docker build - error getting credentials

### ISSUE THREAD

**Issue by chilippso** [docker/for-win#11261]
- [x] I have tried with the latest version of Docker Desktop
  - [x] I have tried disabling enabled experimental features
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: A5DA81D0-A2FF-424C-BF0F-B77C182B756F/20210520130721

### Actual behavior
On new installed system/device (**non-English localized** Windows, WSL2, Docker Desktop, **no credentials in windows credential store**), `docker build .` exits for a fairly simple `Dockerfile` (see: steps to reproduce) with the error:

```
failed to solve with frontend dockerfile.v0: failed to create LLB definition: rpc error: code = Unknown desc = error getting credentials - err: exit status 1, out: `error getting credentials - err: exit status 1, out: `Element nicht gefunden.``
```

### Expected behavior
On new installed system/device (**non-English localized** Windows, WSL2, Docker Desktop, **no credentials in windows credential store**), `docker build .` for a fairly simple `Dockerfile` (see: steps to reproduce) just works as expected.

### Information
  - Is it reproducible? **Yes**
  - Is the problem new? **Yes**
  - Did the problem appear with an update? **No**

  - Windows Version: **20H2 (Build 19042.508)** - non-English localized
  - Docker Desktop Version: **3.3.3 (64133)**
  - WSL2 or Hyper-V backend? **WSL2**
  - Are you running inside a virtualized Windows e.g. on a cloud server or a VM? **No**

### Steps to reproduce the behavior
  1. Install a new copy / set you current copy of Windows 10 Version 20H2 (Build 19042.508) with language set to anything but English
  2. Install WSL2 following the manual at Microsoft Docs with latest Debian image
  3. Install Docker Desktop for Windows Version 3.3.3 (64133), enable WSL integration
  4. Ensure that there are non entries in windows credential manager
  4. Create a (temporary) directory for testing docker build (whether within WSL or Windows filesystem - does not make a difference)
  5. `cd` into the previously created (temporary) testing directory
  6. Create this fairly simple Dockerfile:
```
FROM alpine:3.13.2
```
  7. Issue `docker build .`

Command output:

```
[+] Building 5.3s (3/3) FINISHED
 => [internal] load build definition from Dockerfile                                                               0.1s
 => => transferring dockerfile: 55B                                                                                0.0s
 => [internal] load .dockerignore                                                                                  0.0s
 => => transferring context: 2B                                                                                    0.0s
 => ERROR [internal] load metadata for docker.io/library/alpine:3.13.2                                             5.2s
------
 > [internal] load metadata for docker.io/library/alpine:3.13.2:
------
failed to solve with frontend dockerfile.v0: failed to create LLB definition: rpc error: code = Unknown desc = error getting credentials - err: exit status 1, out: `error getting credentials - err: exit status 1, out: `Element nicht gefunden.``
```

### Working Quick-Fix
I found out that disabling the `buildkit` feature by configuring the Docker Engine via Docker Desktop solves this issue.

```
"features": {
    "buildkit": false
  }
``` 

Nevertheless I wanted to report my struggle here, since I was just installing everything at current versions on a clean and new system and it does not work without custom configuration. Also I do not know the side-effects for disabling `buildkit`, so I am looking for a clean solution, i.e. bugfix, for this basic use-case issue.

### Some guesses
This issue might be related to the WSL users Docker `.config` file.
```
$ cat ~/.docker/config.json
{
  "credsStore": "desktop.exe"
}
```
But maybe I am wrong with this guess.

**Comment by chilippso** [comment]
I just found out that executing `C:\Program Files\Docker\Docker\resources\bin\docker-credential-desktop.exe list` from within standard Windows command prompt gives the same error:

```
error listing credentials - err: exit status 1, out: `Element nicht gefunden.`
```

Credential Helper version: 0.6.3

By the way, editing `~/.docker/config.json` will not last. `"credsStore": "desktop.exe"` gets added automatically on Docker restart.

**Comment by chilippso** [comment]
I investigated a bit further and found something interesting.

First, the issue does not show on English-language systems or when there is any credential saved in windows credential store (It does not matter which one but there has to be one).

So this issue only shows on non-English systems with no credentials stored in the windows credential store.

And here is the explanation:

The currently most recent (released) `Credential Helper` (version 0.6.3) has been released on 16 Jul 2019.

`github.com/docker/docker-credential-helpers` (`docker-credential-wincred.exe`) [depends on](https://github.com/docker/docker-credential-helpers/blob/54f0238b6bf101fc3ad3b34114cb5520beb562f5/wincred/wincred_windows.go#L8) `github.com/danieljoos/wincred`, which got the fix for this issue with the [commit on 13 May 2020](https://github.com/danieljoos/wincred/commit/a594280e412393fe8dcc60cde2d3fc2a6557add7).

So basically `github.com/danieljoos/wincred` in the version used for `github.com/docker/docker-credential-helpers` version 0.6.3 had a "bug" with interpreting (localized) error-strings instead of error-codes. This leads to wrong behavior on non-English systemsn when there are no credentials saved: instead of returning an empty list it returns/"throws" a "false-negative" error.

I tried recompiling `github.com/docker/docker-credential-helpers` (`docker-credential-wincred.exe`) on my own using `github.com/danieljoos/wincred` version `1.1.0` and replaced the binary `docker-credential-wincred.exe` in my docker installation.

Now everything works as expected, without touching docker configuration (i.e. setting `buildkit: false`).

Also note that the current "master" branch for `github.com/docker/docker-credential-helpers` already integrates `github.com/danieljoos/wincred` in version `1.1.0`.

So it is time for a new release guys! I will also head over to `github.com/docker/docker-credential-helpers` and create a linked issue there.

**Comment by stephen-turner** [comment]
This is really helpful, @chilippso, thanks for the great detective work.

**Comment by chilippso** [comment]
You're welcome - I am looking forward for the update 😀

**Comment by Makoehle** [comment]
@chilippso I had the same issue. A work arround was to manually pull the images instead of having the build process doing it. In your case

- `docker pull alpine:3.13.2`

Another thing that might help is to `docker login`.

**Comment by chilippso** [comment]
Thank you @Makoehle for your suggestions but I already identified and solved the issue by building the most recent [docker-credential-helpers](https://github.com/docker/docker-credential-helpers) on my own and replacing the shipped binary.

I will leave this issue open until the patched version of the [docker-credential-helpers](https://github.com/docker/docker-credential-helpers) has been released and deployed by docker-desktop.

Until then, if someone else is affected by this issue, feel free to follow my approach to rebuild the most recent [docker-credential-helpers](https://github.com/docker/docker-credential-helpers) on your own and replace the shipped binary with your custom compiled one.

**Comment by ebriney** [comment]
@chilippso : We are now rebuilding the credential helpers when packaging Docker Desktop.
I used latest master commit in this build: https://desktop-stage.docker.com/win/stable/amd64/64944/Docker%20Desktop%20Installer.exe
Can you try it?

**Comment by chilippso** [comment]
@ebriney: This build works out of the box after installing and fixes this issue.

However, `About Docker Desktop` still states `Credential Helpers: 0.6.3` - so for the _final non-staging release_ you better consider [triggering a new release for the `docker-credential-helpers` Repository](https://github.com/docker/docker-credential-helpers/issues/202) or - if this is not an option - maybe state the git commit hash used during packaging / building for reference.

I am closing this topic now, as it seems this issue will be patched with release version 3.4.0.

**Comment by chilippso** [comment]
So, I don't see any reference in the Change-Log for 3.4.0 so I reopen this issue again and leave it open until closed by someone of the docker team when finally integrated ...

**Comment by stephen-turner** [comment]
I thought this was fixed in 3.4.0, have you tested it? We might have just missed it from the release notes.

**Comment by chilippso** [comment]
@stephen-turner  Sure, I tested the custom build provided by @ebriney but I have not yet tested the finally released Windows version of Docker Desktop.

But the MacOS version states "Credential Helpers: 0.6.4" so I guess it's all ok.

I just read the release notes for 3.4.0 and they don't state this issue so I thought maybe it was shifted to a following patch release (since Credential Helpers 0.6.4 got released just 2 days before Docker Desktop 3.4.0).

I will actually verify the official release 3.4.0 again and report the outcome. But I assume it's all fine.

Cheers,
chilippso

**Comment by geerteltink** [comment]
I just ran into this issue after updating to 3.4.0

```
 => ERROR [php internal] load metadata for registry
------
 > [php internal] load metadata for ...:
------
failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to create LLB definition: failed to authorize: failed to fetch anonymous token: unexpected status: 403 Forbidden
```

**Comment by geerteltink** [comment]
I tracked this down to the beta docker compose v2 version in 3.4.0. When disabling it, it worked again.

I use a registry mirror. Maybe it helps.

**Comment by chilippso** [comment]
@geerteltink I don't think your issue is somehow related to this one here and I guess you better open a new issue for that.

@stephen-turner This issue here is indeed fixed with version 3.4.0 (verified just right now)
